Test TestExistingUpload depends on running in a git working-copy, dirties .git/lfs/objects

[TestExistingUpload](https://github.com/github/git-lfs/blob/master/lfs/upload_test.go#L16) is working with local storage, and is depending on running within a git repository.  This prevents building from a source-tarball (#1201).

It is also then polluting the local git-lfs working-copy's Git LFS object store with a test-artifact in the form of `.git/lfs/objects/98/88/988881adc9fc3655077dc2d4d757d480b5ea0e11`.

Build failure in from-tarball-source build:

```
=== RUN   TestSSHGetExeAndArgsTortoisePlinkCustomPort
--- PASS: TestSSHGetExeAndArgsTortoisePlinkCustomPort (0.00s)
=== RUN   TestExistingUpload
--- FAIL: TestExistingUpload (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x6fb88d]

goroutine 281 [running]:
panic(0x8cf640, 0xc82000e170)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
testing.tRunner.func1(0xc8201ea750)
	/usr/local/go/src/testing/testing.go:467 +0x192
panic(0x8cf640, 0xc82000e170)
	/usr/local/go/src/runtime/panic.go:426 +0x4e9
github.com/github/git-lfs/localstorage.localObjectDir(0x0, 0x9ec1a0, 0x28, 0x0, 0x0)
	<redacted>/go/src/github.com/github/tmp/git-lfs/obj-x86_64-linux-gnu/src/github.com/github/git-lfs/localstorage/localstorage.go:57 +0x19d
github.com/github/git-lfs/localstorage.(*LocalStorage).BuildObjectPath(0x0, 0x9ec1a0, 0x28, 0x0, 0x0, 0x0, 0x0)
	<redacted>/go/src/github.com/github/tmp/git-lfs/obj-x86_64-linux-gnu/src/github.com/github/git-lfs/localstorage/localstorage.go:48 +0x6e
github.com/github/git-lfs/lfs.LocalMediaPath(0x9ec1a0, 0x28, 0x0, 0x0, 0x0, 0x0)
	<redacted>/go/src/github.com/github/tmp/git-lfs/obj-x86_64-linux-gnu/src/github.com/github/git-lfs/lfs/lfs.go:57 +0x4f
github.com/github/git-lfs/lfs.TestExistingUpload(0xc8201ea750)
	<redacted>/go/src/github.com/github/tmp/git-lfs/obj-x86_64-linux-gnu/src/github.com/github/git-lfs/lfs/upload_test.go:109 +0x4f5
testing.tRunner(0xc8201ea750, 0xdc03a0)
	/usr/local/go/src/testing/testing.go:473 +0x98
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:582 +0x892
FAIL	github.com/github/git-lfs/lfs	0.112s
?   	github.com/github/git-lfs/localstorage	[no test files]
?   	github.com/github/git-lfs/script	[no test files]
?   	github.com/github/git-lfs/subprocess	[no test files]
dh_auto_test: go test -v github.com/github/git-lfs github.com/github/git-lfs/commands github.com/github/git-lfs/docs/man github.com/github/git-lfs/git github.com/github/git-lfs/lfs github.com/github/git-lfs/localstorage github.com/github/git-lfs/script github.com/github/git-lfs/subprocess returned exit code 1
make[1]: *** [override_dh_auto_test] Error 1
make[1]: Leaving directory `<redacted>/go/src/github.com/github/tmp/git-lfs'
make: *** [build] Error 2
dpkg-buildpackage: error: debian/rules build gave error exit status 2
```

After running in a proper Git LFS working-copy:

```
cat .git/lfs/objects/98/88/988881adc9fc3655077dc2d4d757d480b5ea0e11 
test
```

Reference commit efdcb06b856b248c99c9374b835f9331ff3ba97b .

--- a/git/git.go
+++ b/git/git.go
@@ -184,9 +184,13 @@ func LocalRefs() ([]*Ref, error) {
 	if err != nil {
 		return nil, fmt.Errorf("Failed to call git show-ref: %v", err)
 	}
-	cmd.Start()
 
 	var refs []*Ref
+
+	if err := cmd.Start(); err != nil {
+		return refs, err
+	}
+
 	scanner := bufio.NewScanner(outp)
 	for scanner.Scan() {
 		line := strings.TrimSpace(scanner.Text())

--- a/git/git_test.go
+++ b/git/git_test.go
@@ -275,6 +275,13 @@ func TestVersionCompare(t *testing.T) {
 }
 
 func TestGitAndRootDirs(t *testing.T) {
+	repo := test.NewRepo(t)
+	repo.Pushd()
+	defer func() {
+		repo.Popd()
+		repo.Cleanup()
+	}()
+
 	git, root, err := GitAndRootDirs()
 	if err != nil {
 		t.Fatal(err)
@@ -380,12 +387,38 @@ func TestGetTrackedFiles(t *testing.T) {
 }
 
 func TestLocalRefs(t *testing.T) {
+	repo := test.NewRepo(t)
+	repo.Pushd()
+	defer func() {
+		repo.Popd()
+		repo.Cleanup()
+	}()
+
+	repo.AddCommits([]*test.CommitInput{
+		{
+			Files: []*test.FileInput{
+				{Filename: "file1.txt", Size: 20},
+			},
+		},
+		{
+			NewBranch:      "branch",
+			ParentBranches: []string{"master"},
+			Files: []*test.FileInput{
+				{Filename: "file1.txt", Size: 20},
+			},
+		},
+	})
+
+	test.RunGitCommand(t, true, "tag", "v1")
+
 	refs, err := LocalRefs()
 	if err != nil {
 		t.Fatal(err)
 	}
 
+	actual := make(map[string]bool)
 	for _, r := range refs {
+		t.Logf("REF: %s", r.Name)
 		switch r.Type {
 		case RefTypeHEAD:
 			t.Errorf("Local HEAD ref: %v", r)
@@ -393,6 +426,22 @@ func TestLocalRefs(t *testing.T) {
 			t.Errorf("Stash or unknown ref: %v", r)
 		case RefTypeRemoteBranch, RefTypeRemoteTag:
 			t.Errorf("Remote ref: %v", r)
+		default:
+			actual[r.Name] = true
 		}
+	}
+
+	expected := []string{"master", "branch", "v1"}
+	found := 0
+	for _, refname := range expected {
+		if actual[refname] {
+			found += 1
+		} else {
+			t.Errorf("could not find ref %q", refname)
+		}
+	}
+
+	if found != len(expected) {
+		t.Errorf("Unexpected local refs: %v", actual)
 	}
 }

